Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,525 @@
+From b73f2b42e37db856d08d557aef374ac6ec51e39f Mon Sep 17 00:00:00 2001
+From: slicklibro <slicklibro@gmail.com>
+Date: Fri, 20 Sep 2019 17:16:14 +1200
+Subject: [PATCH] Ticket level per-player view distance API
+
+This PR implements the per-player view distance API at the ticket level.
+This somewhat clean solution comes from the result of many complicated ones (and research).
+The view distance will determine how far the player can add tickets to chunks.
+The tracking view distance will determine how far the player can see/load chunks with tickets.
+All chunks with player tickets within this range will also be served by the server, allowing shared view distance from other players without impacting performance.
+
+diff --git a/src/main/java/net/minecraft/server/ChunkMapDistance.java b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+index 63a688725..4db2e954f 100644
+--- a/src/main/java/net/minecraft/server/ChunkMapDistance.java
++++ b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+@@ -43,6 +43,16 @@ public abstract class ChunkMapDistance {
+     private final LongSet l = new LongOpenHashSet();
+     private final Executor m;
+     private long currentTick;
++    //Paper start ticket level pp-view-dist
++    protected final it.unimi.dsi.fastutil.ints.Int2ObjectMap<Square> squares = new it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<>();
++    protected it.unimi.dsi.fastutil.longs.Long2ByteMap update = new it.unimi.dsi.fastutil.longs.Long2ByteOpenHashMap();
++    class Square {
++        public int x, y, u, v, ticket;
```

Comment:
these variable names are short and undescriptive of what they are
this goes for the rest of this class, stuff like this is unacceptable:

1. `Square sto = ...`
2. `if(sqr.ticket == -2) {`
3. `byte f = ...`

along with many other very poor names for variables and magic constants you have used

We need to maintain these patches across updates, fix issues, etc and right now these names are really going to make that difficult

Reply:
most of these names are standard placehodlers for logic within the alg, x, y ,u, v are relatively self explanatory to represent corners of a rectangle though we can change sqr.ticket, and some in between var names

Reply:
considering that `y` refers to the lower z coordinate, no they're not self-explanatory to anybody but you

Reply:
formed for a 2d plane will change that though...

Reply:
any other nitpicks about var names that you would like changed, a list would be nice

Reply:
most of these are just algebraic placeholders for the alg as a note

Reply:
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R45
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R55
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R59
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R69
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R107
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R115
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R170
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R173
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R184
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R186
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R228
- https://github.com/PaperMC/Paper/pull/2572/files#diff-55ef5c1a62185e284c903fc1d128fc97R238

This list also includes the usages of magic constants, please write them to a static final field with its own descriptive name

Reply:
changed, will be pushed on next commit